## COSC 1210
### print(), input(), value storage and chaos

Objectives
- use the chaos function to illustrate the importance of value storage

In [6]:

def main():
    a = 1/7
    b = a+a+a+a+a+a+a
    print(b)
main()

0.9999999999999998


##### Explain why the function above provides an answer that is different than the algebraic solution
The computer can not store an infinte number.

In [11]:
def main():
    a = 1/7
    b = a * 7
    c = a + a + a + a + a + a + a
    print("b is", b)
    print("c is", c)
    print(b-c)
main()

b is 1.0
c is 0.9999999999999998
2.220446049250313e-16


##### Explain why b and c provide different answers
Because for b it has built in rounding, but adding doesnt round as early as addition


##### How large is the difference between c and b?
2.220446049250313e-16



In [18]:
def main():
    print("This program illustrates a chaotic function")
    iterations = 100
    x1 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
    print("For", iterations, "iterations, x1 is:", x1)
main()

This program illustrates a chaotic function
For 100 iterations, x1 is: 0.5126527786207501


##### If you calculate x1 for 10 iterations vs 100 iterations, what answers do you get?

for 10 iterations: 0.9735732406265619

for 100 iterations: 0.5126527786207501

##### Why are these different?
they are different because the code was repeated 90 more times making the number smaller, the number changes each time through the calculation


In [21]:
def main():
    print("This program illustrates a chaotic function")
    iterations = 100
    x1 = 0.15
    x2 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
        x2 = 3.9 * (x2 - x2 * x2)
    print("For", iterations, "iternations, x1 is now", x1, "and x2 is now", x2)
main()

This program illustrates a chaotic function
For 100 iternations, x1 is now 0.5126527786207501 and x2 is now 0.16411537941360743


##### If you calculate x1 and x2 for 10 iterations vs 100 iterations, what answers do you get?
x1 10 iterations: 0.9735732406265619
x2 10 iterations: 0.9735732406265659

x1 100 iterations: 0.5126527786207501
x2 100 iterations: 0.16411537941360743

Why are these different?
    These are different because in the different equations they use a different amount of variables, causing it to round different, the less variable use the more accurate that the answer is going to be. The rounding causes the numbers to be completely different when you have 100 iterations becasue the rounding beig changed slightly every time throughout the 00 times causes the numbers to be completely different.

In [24]:
### Rewrite the chaos function so that the use can enter x
def main():
    print("This program illustrates a chaotic function")
    iterations = 100
    x1 = 0.15
    x2 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
        x2 = 3.9 * (x2 - x2 * x2)
    print("For", iterations, "iternations, x1 is now", x1, "and x2 is now", x2)
main()

This program illustrates a chaotic function
For 100 iternations, x1 is now 0.5126527786207501 and x2 is now 0.16411537941360743
